compiler/src/iree/compiler/Codegen/Common/CPU/CPUMaterializeEncodingPass.cpp

```diff
@@ -193,8 +193,9 @@ enumerateMatmulTileX86_64(EncodingUser user, TypeRange elementTypes,
     };
   }
 
-  if (lhs.isSignlessInteger(8) && rhs.isSignlessInteger(8) &&
-      out.isSignlessInteger(32)) {
+  if (out.isSignlessInteger(32) &&
+      ((lhs.isSignlessInteger(8) && rhs.isSignlessInteger(8)) ||
+       (lhs.isSignlessInteger(16) && rhs.isSignlessInteger(16)))) {
     if (hasFeature(target, "+avx512vnni")) {
       // This is the same tile size as with VPMADDWD as the only difference
       // is that VPDPWSSD accumulates. VPDPBUSD would call for {16, 16, 4} but
```

Comment:
Out of curiosity, why do we tile `M` and `N` to <= 16 for `i8` cases? Wouldn't this not make use of the full 512 bits (`16*2*8=256`)?

Reply:
Because on x86, the (signed x signed) `i8` cases have to sign-extend both LHS and RHS to `i16` to feed to VPMADDWD/VPDPWSSD.  x86 just does not have great support for `i8 x i8 -> i32`.  As the comment notes, the VNNI instruction that we're supposed to be using for that, VPDPBUSD, supports only the (unsigned x signed) case, and that makes it difficult to use here.